mbedtls: remove unused configuration header files

valeriosetti opened the pull request:
- Remove all unused configuration header files. These were likely copied from Mbed TLS in the past, but not updated afterwards. At the same time they are not used in the Zephyr's scope, so it's useless to keep them around.

- Do a bit of renaming:
    - `config-tls-generic.h` -> `config-mbedtls.h`
    - `Kconfig.tls-generic` -> `Kconfig.mbedtls`

valeriosetti committed:
mbedtls: remove unnecessary headers

These headers are basically a copy of those found in the official
Mbed TLS repo in the "configs" folder. However these are unmaintaned
(i.e. not updated when Mbed TLS is bumped to a new version) and
also unused in Zephyr. Therefore there's no benefit in keeping
them around.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

tomi-font commented on modules/mbedtls/Kconfig.extra:
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) 2024 BayLibre SAS
+# SPDX-License-Identifier: Apache-2.0
+
+# Extra options the are not strictly necessary for Mbed TLS itself, but that
+# configure it properly for other modules/subsystems.
```
Actually, I am not sure about this file. We still have `Kconfig` which contains a bunch of Kconfig options and they don't all directly translate to Mbed TLS configuration items. Though they are not so tightly tied to other parts of Zephyr. Maybe for now we could have what's here in `Kconfig`? Or do you see this file growing?

tomi-font commented on modules/mbedtls/Kconfig.tls:
Wondering whether this file should include `x509` in its name as well...

tomi-font commented on modules/mbedtls/Kconfig.tls:
```diff
@@ -0,0 +1,222 @@
+
+endif # MBEDTLS_TLS_VERSION_1_2 || MBEDTLS_TLS_VERSION_1_3
+
+menu "Ciphersuites"
```
Isn't this key exchanges rather than ciphersuites?

valeriosetti replied:
well, all the TLS 1.2 related entries had "ciphersuite" in the prompt description even though the name of the Kconfig symbol is "_KEY_EXCHANGE_". I suspect this somehow relates to the fact that in TLS 1.2 also includes details about other allowed features (ex: signatures, etc), whereas in TLS 1.3 you really only specify the key exchange type. Just guessing though, so if you prefer me to fix this I'll do. Just let me know ;)

tomi-font commented on modules/mbedtls/Kconfig.crypto:
```diff
@@ -285,23 +267,9 @@ config MBEDTLS_ENTROPY_POLL_ZEPHYR
-	select MBEDTLS_MD
-	help
-	  Enable PKCS5 functions
+menu "PSA crypto core"
```
PSA **C**rypto

tomi-font commented on modules/mbedtls/Kconfig.crypto:
```diff
@@ -3,32 +3,22 @@
 # Mbed TLS's crypto library options
 
-menu "Ciphersuite configuration"
+menu "Elliptic curves algorithms and groups"
```
`curve~~s~~`

tomi-font commented on modules/mbedtls/Kconfig.crypto:
```diff
@@ -89,11 +79,13 @@ config MBEDTLS_ECP_DP_CURVE448_ENABLED
+endif # MBEDTLS_ECP_C
 
-endif
+endmenu # Elliptic curve libraries
```
mismatch with opening menu

tomi-font commented on modules/mbedtls/Kconfig.crypto:
```diff
@@ -371,12 +339,6 @@ config MBEDTLS_PSA_CRYPTO_CLIENT
 	depends on BUILD_WITH_TFM || MBEDTLS_PSA_CRYPTO_C
```
Actually, how about we remove this Kconfig option and only use `PSA_CRYPTO_CLIENT`? They're sort of duplicates and `CONFIG_MBEDTLS_PSA_CRYPTO_CLIENT` already doesn't map to a single `#define` in the Mbed TLS configuration file.

tomi-font commented on modules/mbedtls/Kconfig.crypto:
```diff
@@ -393,12 +355,51 @@ config MBEDTLS_PSA_P256M_DRIVER_RAW
+endmenu # PSA crypto core
+
+config MBEDTLS_HKDF_C
+	bool "HMAC-based Extract-and-Expand Key Derivation Function"
```
Why specifying `Extract-and-Expand` here?

tomi-font commented on modules/mbedtls/Kconfig.crypto:
```diff
@@ -393,12 +355,51 @@ config MBEDTLS_PSA_P256M_DRIVER_RAW
+	bool "HMAC-based Extract-and-Expand Key Derivation Function"
+
+config MBEDTLS_GENPRIME_ENABLED
+	bool "prime-number generation code."
```
`Prime number generation`

tomi-font commented on modules/mbedtls/Kconfig.crypto:
```diff
@@ -393,12 +355,51 @@ config MBEDTLS_PSA_P256M_DRIVER_RAW
+	bool "Password-based encryption functions"
+	select MBEDTLS_MD
+	help
+	  Enable PKCS5 functions
```
This could be removed.

tomi-font commented on modules/mbedtls/configs/config-mbedtls.h:
```diff
@@ -11,14 +12,15 @@
 #define MBEDTLS_CONFIG_H
 
-/* System support */
+/* Always enabled build symbols */
```
minor: `Always-enabled`

tomi-font commented on modules/mbedtls/configs/config-mbedtls.h:
```diff
@@ -11,14 +12,15 @@
 #define MBEDTLS_PLATFORM_EXIT_ALT
 #define MBEDTLS_NO_PLATFORM_ENTROPY
 
+/* Mapping of Kconfig options to Mbed TLS build symbols */
```
Suggest leaving a newline afterwards.

tomi-font commented on modules/mbedtls/Kconfig.crypto:
```diff
@@ -85,6 +85,26 @@ endif # MBEDTLS_ECP_C
+	bool "PKCS1 v2.1 (OAEP and PSS)"
+
+config MBEDTLS_GENPRIME_ENABLED
+	bool "Prime-number generation"
```
`Prime number`

tomi-font commented on modules/mbedtls/Kconfig.tls:
```diff
@@ -62,6 +62,9 @@ config MBEDTLS_KEY_EXCHANGE_ECDHE_PSK_ENABLED
 
 config MBEDTLS_KEY_EXCHANGE_RSA_PSK_ENABLED
 	bool "RSA-PSK based ciphersuite modes"
+	imply MBEDTLS_RSA_C if !PSA_CRYPTO_CLIENT
```
Why not `select`? And what's the reasoning around `PSA_CRYPTO_CLIENT`?
+ IMO we could define a bundle that enables those 3 Kconfig option as this is done several times below.

valeriosetti replied:
> And what's the reasoning around PSA_CRYPTO_CLIENT?

The idea was to select RSA_C (i.e. the legacy solution) only if there is not a PSA support for PSA available. But now that you made me notice this, I might have used PSA_WANT symbols instead here

tomi-font reviewed (changes requested):
A very welcome rework!

valeriosetti commented on modules/mbedtls/Kconfig.tls:
```diff
@@ -0,0 +1,192 @@
+	  which allows to identify DTLS connections across changes
+	  in the underlying transport.
+
+config MBEDTLS_SSL_MAX_CONTENT_LEN
```
Actually the entire `Kconfig.tls` is gated by `MBEDTLS_BUILTIN` in the main `Kconfig`

valeriosetti committed:
mbedtls: rename tls-generic Kconfig and configuration header file

Since there's now a single Kconfig and header file for Mbed TLS it makes
more sense to name them:
- Kconfig.mbedtls (instead of Kconfig.tls-generic)
- config-mbedtls.h (instead config-tls-generic.h)

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented:
Since it was a bit hard to get reviews on this, I strongly reduced the scope of the PR. Now it just removes unused files and renames what's left for Mbed TLS.

valeriosetti commented on west.yml:
```diff
@@ -303,7 +303,7 @@ manifest:
         - crypto
     - name: mcuboot
-      revision: 346f7374ff4467e40b5594658f8ac67a5e9813c9
+      revision: pull/124/head
```
Yes, the reference to https://github.com/zephyrproject-rtos/mcuboot/pull/124 is surely not correct now, but I don´t know if I can directly point to https://github.com/mcu-tools/mcuboot/pull/2222. While creating PR 2222 I noticed that the `main` branch of https://github.com/mcu-tools/mcuboot is a bit newer than https://github.com/zephyrproject-rtos/mcuboot, so I guess that the process is:
- review & merge https://github.com/mcu-tools/mcuboot/pull/2222
- create a PR in https://github.com/zephyrproject-rtos/mcuboot which cherry-picks my fix
- update this PR with the PR in https://github.com/zephyrproject-rtos/mcuboot
Am I wrong?

de-nordic replied:
> 
>     * create a PR in https://github.com/zephyrproject-rtos/mcuboot which cherry-picks my fix
> 

No.

> Am I wrong?

Yes ;)

We sync mcuboot to have exactly the same history as upstream, so that is only done via full main branch syncs.
Like here:
https://github.com/zephyrproject-rtos/zephyr/pull/85671

valeriosetti commented:
I just removed the manifest update for mcuboot because https://github.com/mcu-tools/mcuboot/pull/2222 has been merged so now it's only matter to wait for https://github.com/zephyrproject-rtos/mcuboot to be updated with the latest change from upstream and the CI of this PR will be green again.

tomi-font commented:
@d3zd3z @ceolin please review

valeriosetti commented:
Rebase done without any code change. The PR is ready for reviews again.

dkalowsk reviewed (approved):
Thanks for this clean up and making it easier for end users.

valeriosetti commented:
> @valeriosetti this will need rebasing - thx!

@kartben rebase done, CI green and already got some approvals. Can you please merge this before any conflict pops up?

tomi-font commented:
> @kartben rebase done, CI green and already got some approvals. Can you please merge this before any conflict pops up?

This is still missing approval from one of the assignees. @ceolin @d3zd3z please review
